### ISSUE TITLE
Poor performance in haproxy 2.0+

### ISSUE THREAD

**Issue by ngaugler** [haproxy/haproxy#822]
## Output of `haproxy -vv` and `uname -a`
```
HA-Proxy version 2.0.17-1ppa1~xenial 2020/08/01 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=40).
Built with OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
Running on OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.1
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.21 2016-01-12
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
        prometheus-exporter

Available filters :
        [SPOE] spoe
        [COMP] compression
        [CACHE] cache
        [TRACE] trace
```

```Linux 4.4.0-141-generic #167-Ubuntu SMP Wed Dec 5 10:40:15 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux```

## What's the configuration?

```
global
        log /dev/log    local0 notice
        chroot /var/lib/haproxy

        stats socket /run/haproxy/haproxy_20.sock mode 664 level admin
        stats timeout 30s

        user haproxy
        group haproxy
        daemon

        nbproc 2
        nbthread 1

        maxconn 500000


defaults
        log     global
        mode    http

        option  dontlognull
        option  dontlog-normal
        option  redispatch

        option  tcp-smart-accept
        option  tcp-smart-connect

        timeout connect 2s
        timeout client  50s
        timeout server  50s
        timeout client-fin 1s
        timeout server-fin 1s

        maxconn 150000

        errorfile 400 /etc/haproxy/errors/400.http
        errorfile 403 /etc/haproxy/errors/403.http
        errorfile 408 /etc/haproxy/errors/408.http
        errorfile 500 /etc/haproxy/errors/500.http
        errorfile 502 /etc/haproxy/errors/502.http
        errorfile 503 /etc/haproxy/errors/503.http
        errorfile 504 /etc/haproxy/errors/504.http

listen bidder
        bind    10.0.0.1:80    defer-accept process 1
        bind    10.0.0.2:80    defer-accept process 1
        bind    10.0.0.3:80    defer-accept process 1
        bind    10.0.0.4:80    defer-accept process 1
        bind    10.0.0.5:80    defer-accept process 1
        bind    10.0.0.1:80    defer-accept process 2
        bind    10.0.0.2:80    defer-accept process 2
        bind    10.0.0.3:80    defer-accept process 2
        bind    10.0.0.4:80    defer-accept process 2
        bind    10.0.0.5:80    defer-accept process 2
        mode    http

        maxconn 65000
        fullconn 65000

        balance leastconn
        http-reuse safe

        server  s001    10.0.1.1:80   weight 100 source "10.0.0.250"
        server  s002    10.0.1.2:80   weight 100 source "10.0.0.251"
        server  s003    10.0.1.3:80   weight 100 source "10.0.0.252"
        server  s004    10.0.1.4:80   weight 100 source "10.0.0.253"
        server  s005    10.0.1.5:80   weight 100 source "10.0.0.250"
        server  s006    10.0.1.6:80   weight 100 source "10.0.0.251"
        server  s007    10.0.1.7:80   weight 100 source "10.0.0.252"
        server  s008    10.0.1.8:80   weight 100 source "10.0.0.253"
        server  s009    10.0.1.9:80   weight 100 source "10.0.0.250"
        server  s010    10.0.1.10:80  weight 100 source "10.0.0.251"
        server  s011    10.0.1.11:80  weight 100 source "10.0.0.252"
        server  s012    10.0.1.12:80  weight 100 source "10.0.0.253"
        server  s013    10.0.1.13:80  weight 100 source "10.0.0.250"
        server  s013    10.0.1.14:80  weight 100 source "10.0.0.251"
```

On haproxy 1.6 the configs are much more complicated with stats listeners, health checks and agent checks.  The more complicated config was originally tested on haproxy 2.0.2, but it was simplified to attempt to narrow down what was the actual problem.  The actual problem appears to be in the core of the code base.

## Steps to reproduce the behavior

1.  Connect ~1000 clients
2.  Send requests through that take ~100-150ms to process
3.  Once the request is complete, do not close the socket, send again.  
4.  Repeat steps 2-3.

## Actual behavior

When running haproxy2.0, or even 2.2.2, the CPU usage is significantly higher than when using haproxy1.6.

Please see the following CPU usage from top from a Saturday (lower traffic):
```
195277 haproxy   20   0  163592  81580   3016 S  56.2  0.1   0:07.23 haproxy-2.0
195276 haproxy   20   0  163920  81848   3056 S  50.0  0.1   0:08.04 haproxy-2.0
```
Attempt to run haproxy 2.2.2:
```
195039 haproxy   20   0  159852  75164   3476 R  45.0  0.1   0:18.01 haproxy-2.2
195040 haproxy   20   0  160020  75620   3480 S  45.0  0.1   0:18.71 haproxy-2.2
```

## Expected behavior

CPU usage to be consistent or less than the previous code base.

```
195572 haproxy   20   0  524540  21072   1988 R  28.5  0.0   0:08.49 haproxy-1.6
195592 haproxy   20   0  525296  37236  17436 S  28.1  0.1   0:08.34 haproxy-1.6
```

## Do you have any idea what may have caused this?

I suspect this is related to the new core code written in 2.0.

## Do you have an idea how to solve the issue?

It does not appear to be related to system calls to the kernel.  I believe it to be some other inefficiency within the code base.  I can easily reproduce this as it happens instantly when attempting to run the new code base.  I would presume the developers would need to run a cpu profiler to determine which functions are taking the most cpu, but I am not an avid C programmer.

This discussion started on the mailing lists over a year ago.  The problem still persists and we are still stuck on a legacy version of the code base.  https://discourse.haproxy.org/t/2-0-1-cpu-usage-at-near-100-after-upgrade-from-1-5/4065/27

**Comment by wtarreau** [comment]
The increased CPU usage is expected as it comes from the architectural changes required to support more protocols (like H2), threading, etc. However the peak performance when running with a few threads definitely is higher with 2.2 than 1.6. Regardless, your numbers are socking, you're suggesting 1000 connections and 100-150ms extra response time, this is extreme, and would indicate a perf limit of 7-10000 requests per second, and I don't see how this is possible for now, especially since your CPUs are not even saturated so there's something related to waiting for something. The numbers I'm getting are 10-30 times higher on my laptop. Could you please share some logs which would help figure where the extra time is spent (request, queue, connection, reponse, etc) ?

Your 100-150ms makes me think you might be facing an unexpected 200ms delay in the stack due to an undesired MSG_MORE somewhere. I'm currently looking at another issue (#825) which tends to indicate this might happen on 100-continue for example, so depending on what I find there I may ask you to test one thing or another. In the mean time if you have any logs it would be greatly appreciated.

**Comment by ngaugler** [comment]
Thank you for the prompt response, I apologize for the delay. I am willing to provide any stats and run any experiments you need to further investigate.  I connected a stats listener to each of the forks and grabbed a 'show stat' command from one of the sockets.  It appears my estimate of an average response time of 100-150ms was a bit off.  If there are any other commands you need me to execute, please let me know.

[haproxy-2.2.2-show-stat.txt](https://github.com/haproxy/haproxy/files/5139061/haproxy-2.2.2-show-stat.txt)

I understand there may be overhead with the new architectural changes.  I would hope that it can be made more efficient than a 52% increase in CPU usage.  I also grabbed the top output again today, since traffic is a tad higher than a Saturday.

haproxy-1.6.15
```
195572 haproxy   20   0  537192  33820   2052 S  35.9  0.1   4815:14 haproxy
195592 haproxy   20   0  538036 459272 426600 S  33.6  0.7   4775:28 haproxy
```


haproxy-2.2.2
```
146226 haproxy   20   0  161384  77412   3840 R  54.5  0.1   5:36.83 haproxy-2.2
146227 haproxy   20   0  163400  81536   3716 R  53.2  0.1   5:33.75 haproxy-2.2
```

**Comment by wtarreau** [comment]
So your stats look OK to me. What I'm seeing here is that the average response time on the last 1024 requests is roughly 350 ms. But I really can't see anything wrong there. If you have a way to isolate a set of requests using a special marker (e.g. adding a specific header) and take a capture on both sides, that would help. Even just on the server side as a starter. We could see if a connection is quikly reused or not. Maybe for any reason the server takes time to process a new request over the same connection (e.g. if it has an acceptor thread that dispatches requests to worker threads) ?

One way to check for this hypothesis would be to reduce the number of servers, resulting in an increased load on them, and to see if the delay increases or not. Or conversely if you have some spare servers that you can throw into the farm to see if delays decrease, that would indicate the same. For example I remember a long time ago having seen servers issuing DNS resolutions of incoming requests' source address to enrich logs, but that came with a huge cost!

**Comment by ngaugler** [comment]
Let me think about that for a bit.  The other end is a highly scaled internal golang project that is doing quite a bit of activity.    Please remember, I can stop haproxy 1.6, start haproxy 2.2, see the problem, stop haproxy 2.2, start haproxy 1.6 and things are instantly fine. I can continue this all day long and never experience issues with haproxy 1.6 (or haproxy 1.8 for that matter).

**Comment by wtarreau** [comment]
OK thanks for the info. But then would you be able to capture some traffic so that we can analyse what's happening ?

**Comment by ngaugler** [comment]
I am back, sorry it's been awhile.  We completely removed the need for this type of load balancing from our platform.  It's much more efficient without it.  That being said I am trying to implement 2.2 and the threaded model for frontend load balance (http/https) and again it's horribly inefficient.   With 1.6 running we have 35-40% idle cpu.  With haproxy 2.2 we have 4% idle CPU and connections are being dropped.  How is anyone running this software in production with real work loads when it uses 50% more CPU than the previous versions?  What's the best step to take from here to figure out how to solve this?

**Comment by wtarreau** [comment]
Actually that doesn't correspond to many others' observations, this is why others are not running into such trouble. The increase of CPU usage from 1.6 to 2.2 was measured to be around 30% on single-threading, which is attributed to the more flexible architecture allowing to support various protocols (HTTP/2, FCGI, hopefully QUIC soon), more fairness between tasks, and the threading support. However this is a static cost and you can go way higher as soon as you have more than one thread, which corresponds to virtually all machines where haproxy is deployed nowadays. Just look at this example below where I've run the exact same test on 1.6.15 then on latest 2.4-dev without changing anything to the config nor command line:

```
root@lg34:~# ./bin/h1load -t 16 -c 300 198.19.0.192:8001/
#     time conns tot_conn  tot_req      tot_bytes    err  cps  rps  Bps  bps
1609861003   300      300    97872       12296844      0 300. 97k8 12M2 98M3 
1609861004   300      300   199411       25087986      0 0.00 101k 12M7 102M 
1609861005   300      300   300997       37887822      0 0.00 101k 12M7 102M 
1609861006   300      300   402059       50621634      0 0.00 100k 12M7 101M 
1609861007   300      300   502797       63314622      0 0.00 100k 12M6 101M 
1609861008   300      300   603348       75984048      0 0.00 100k 12M6 101M 
1609861009   300      300   703732       88632558      0 0.00 100k 12M6 101M 
^C
root@lg34:~# ./bin/h1load -t 16 -c 300 198.19.0.192:8001/
#     time conns tot_conn  tot_req      tot_bytes    err  cps  rps  Bps  bps
1609861039   300      300   681093       86462209      0 300. 681k 86M4 691M 
1609861040   300      300  1381300      175388371      0 0.00 699k 88M8 710M 
1609861041   300      300  2080618      264201757      0 0.00 698k 88M7 709M 
1609861042   300      300  2777926      352759873      0 0.00 696k 88M4 707M 
1609861043   300      300  3473941      441153778      0 0.00 695k 88M3 706M 
1609861044   300      300  4167384      529221039      0 0.00 692k 87M9 703M 
1609861045   300      300  4849135      615803416      0 0.00 681k 86M4 691M 
^C

```
It jumped from 101k req/s to 695k req/s on a 8-core machine Xeon. For sure we can probably do better but I'd claim that it scaled extremely well if 2.4 could be **6.9 times faster** than 1.6 on **8 cores**, especially if we consider that we still need some CPU for the network card!

As such, if you manage to put it in situations where it shamefully fails (what you describe above is a failure), please share as many details as possible (config, workload type, observations, logs maybe) and we may also ask you to share some outputs of "show activity", and possibly even to enable profiling and threads debugging. It's perfectly possible you've hit a horrible corner case, maybe one we can address, maybe something that we will not for some time due to various technical limitations, but addressing bottlenecks has always been our main motivation, so we're all interested!

**Comment by ngaugler** [comment]
Unfortunately your whole theory is flawed if you presume that everyone was running 1.6 with one single process.  The most efficient way to make 1.6 utilize all of the available resources of a server was to run multiple processes.  

The use case that caused me to open up this ticket is no longer running in our facilities.  We've merged the two applications into a single application to reduce latency and simplify our scaling, so we no longer need to load balance internally between application A and application B.

The current use case is, what I would consider, the most basic way of scaling haproxy1.6.  We have two frontends, one for http, the other for https, each with 30 different bind statements to provide each process its own socket.  We have a single backend, which load balances all of the actual application servers.  The backend is currently running on 10 different processes (due to the need for more cpu), which is split up with 10 different bind statements so each process has its own socket.  The backend processes listen on unix sockets and accept-proxy.  The two frontends load balance across 10 server lines connecting to the backend via unix sockets.  It scales well, unfortunately there is waste sending requests through the unix sockets and waste with 10 different processes doing health checks to the application servers.  Also fine tuning the balance of frontend vs backend processes is a hassle.  Either way, it's twice as efficient as running haproxy 2.2.  

I can produce conf files.  I can run commands, I can probably even screen share you into the server.  The bottom line is haproxy 2.2 is going to require me to double my hardware just to handle my current work load with 1.6.   

I have tried to sanitize the config files and upload them.   I can provide 'show activity' from various processes (frontend vs backend) on 1.6.  Unfortunately I have to run haproxy 2.2 in the middle of the night when traffic is less than half otherwise I drop traffic.  I am not sure when I will be able to get data out of haproxy 2.2.

[haproxy-1.6-config.txt](https://github.com/haproxy/haproxy/files/5773323/haproxy-1.6-config.txt)
[haproxy-2.2-config.txt](https://github.com/haproxy/haproxy/files/5773324/haproxy-2.2-config.txt)

**Comment by wtarreau** [comment]
Hi,

I never presumed that everyone was running 1.6 with a single process and know pretty well that some were forced to use multi-process with all the problems that came with that mode (inconsistent checks, heavier resources used by health checks on the servers, excessive socket usage on servers due to per-process idle connections, no stick-table synchronization sometimes requiring to chain two layers, no stats aggregation, impossibility to set a meaningful maxconn value, leastconn meaning nothing, DNS resolution making basically no sense, etc). Threads were precisely created so that one doesn't have to choose between performance and manageability.

But I also know that some corner cases used to exist where this was not a problem. And we must do our best to make sure these ones continue to work fine, and ideally with less hardware. And we do know that for a number of users, the hardware size actually reduced after switching to the most modern versions (e.g. less RAM for maps/stick tables, optimal HTTP reuse meaning less connections, etc). In any case if there's something causing a significant perf regression (and half the perf is a huge one) I want to understand the cause.

In your 2.2 config I'm seeing a few things I'm interested in knowing how you migrated them to threads:
  - can you confirm that you moved to a single "bind" line and multiplied its "maxconn" and "backlog"
    parameters by the number of threads so that you get the same total capacity ? Otherwise you could
    end up saturating 30 times faster, and constantly enabling/disabling listening, which would be
    devastating for the performance.

In fact I'm not seeing anything else suspicious. If you confirm the point above, another thing to check would be to run haproxy under "perf record" and then run "perf report" to see where most of the time was spent. Sometimes it already helps spot a dirty issue.

It's just unclear to me whether the process-based 2.2 config you posted currently is the one showing perf issues or the one you need to use to run under similar performance as 1.6. What happens when you run your (simpler and cleaner) 1.6 config on 2.2 ?

One point regarding leastconn is that it focuses on the bet possible accuracy which requires fine synchronization between threads. By just using "balance random" you'll get most of it (it's not just random, it's the popular P2C algo which picks the least loaded of N servers, defaulting to 2, and is very close to leastconn without the synchronization cost; you can even use random(3) to get the best out of 3 servers if you want, but 2 already is quite good and definitely way more accurate than what you used to have with many processes).

thanks!
Willy

**Comment by ngaugler** [comment]
Hello sir,

Thank you for your detailed response.  The configs I pasted on the last message were copied from my 1.6 and 2.2 configs from the server.  I pasted them into a text editor and then sanitized the sensitive information (IP addresses, domain name, etc).  I just reviewed the math and here's what I've come up with:

http_proxy:
backlog  1.6: 30,000  2.2:  32,000
maxconn 1.6: 116,000  2.2:  90,000

https_proxy:
backlog  1.6: 15,000  2.2:  15,000
maxconn 1.6: 22,500 2.2:  30,000

I believe we are under 90,000 on http under normal operation conditions.  That being said, I just ran another test with the following changes:

http_proxy maxconn: 120,000
balance random


I restarted haproxy 1.6 to give it a "clean" start:

```
procs -----------memory---------- ---swap-- -----io---- -system-- ------cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
23  0      0 52528284 754288 6815576    0    0     0     0    0    0 16 26 58  0  0
25  0      0 52529180 754288 6805076    0    0     0     1 1116549 532807 24 40 36  0  0
27  0      0 52504476 754288 6797992    0    0     0     0 1118059 581253 23 38 39  0  0
25  0      0 52526388 754288 6791848    0    0     0     0 1117254 518685 24 41 35  0  0
29  0      0 52242436 754288 6837156    0    0     0     0 1091617 496256 26 41 33  0  0
```


I then restarted with haproxy 2.2 to give it a "clean" start:

```
procs -----------memory---------- ---swap-- -----io---- -system-- ------cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
39  0      0 50187540 754288 6723096    0    0     0     0    0    0 16 26 58  0  0
40  0      0 50065408 754288 6725408    0    0     0     3 1694482 82186 73 22  4  0  0
41  0      0 50133572 754288 6706964    0    0     0     5 1720880 80918 73 23  5  0  0
40  0      0 49993460 754288 6697212    0    0     0     5 1751812 79851 73 23  4  0  0
40  0      0 49989860 754288 6708444    0    0     0     3 1743841 78978 73 23  5  0  0
```

The number of requests processed per second dropped from 155,000 to 33,000 (as reported by our application to influx/grafana not from haproxy stats) and CPU usage went from 61% to 95%.  Restarting with haproxy 1.6 fixed everything.


I will try and grab perf data during an off peak hour when there is less traffic.

**Comment by wtarreau** [comment]
So the test above is with "balance random" if I understand correctly.

What's the problem on this machine ? I'm seeing 40 processes fighting for the run queue, but in your conf you're running only 2 processes or threads. Are you certain it's healthy and that it can guarantee that the CPU haproxy is running on are not constantly monopolized ? Otherwise that could explain a lot of things, because preempting a thread when it holds a lock with has for a side effect to let the other ones wait much longer. This will possibly be visible in your perf data. Thanks!

**Comment by ngaugler** [comment]
Yes, you are correct, the above test was with 'balance random' instead of 'balance leastconn'

I am sorry, but I do not see where I am only running on '2 processes or threads' in the haproxy-2-2-config.txt file.  It is my understanding that haproxy will default the number of threads to the numbers of cores.  Please see the attached haproxy -vv output showing that.

Other than keepalived, salt-minion, and zabbix-agent, haproxy is the only daemon installed and running that is not a base system daemon.  I have 4 identical servers in this facility, 4 on the west coast and 3 in Amsterdam.  All solely devoted to haproxy.  I have attached 3 copies of the output of 'top'.   One for haproxy-1.6, one for haproxy-2.2 and second for haproxy-2.2 with the toggle for 'Threads' enabled. 

I will attempt to port the 1.6 configs to 2.2 so I can run in 2.2 without threads but instead 40 processes.  I will also grab perf data from both 2.2 threaded and 2.2 without threads.
[haproxy-2.2-threads.txt](https://github.com/haproxy/haproxy/files/5871302/haproxy-2.2-threads.txt)
[haproxy-2.2-top.txt](https://github.com/haproxy/haproxy/files/5871303/haproxy-2.2-top.txt)
[haproxy-2.2-vv.txt](https://github.com/haproxy/haproxy/files/5871304/haproxy-2.2-vv.txt)
[haproxy-1.6-top.txt](https://github.com/haproxy/haproxy/files/5871305/haproxy-1.6-top.txt)

**Comment by wtarreau** [comment]
> I am sorry, but I do not see where I am only running on '2 processes or threads' in the haproxy-2-2-config.txt file.

Ah OK, I was basing my analysis on the updated config showing "nbproc 2" above.

> It is my understanding that haproxy will default the number of threads to the numbers of cores.

Yes!

> Please see the attached haproxy -vv output showing that.

OK then all 40 CPU involved in your vmstat output can definitely come from haproxy.

> Other than keepalived, salt-minion, and zabbix-agent, haproxy is the only daemon installed and running that is not a base system daemon. I have 4 identical servers in this facility, 4 on the west coast and 3 in Amsterdam. All solely devoted to haproxy.

This sounds perfect.

> I have attached 3 copies of the output of 'top'. One for haproxy-1.6, one for haproxy-2.2 and second for haproxy-2.2 with the toggle for 'Threads' enabled.

Thanks for this. It is indeed extremely concerning!

> I will attempt to port the 1.6 configs to 2.2 so I can run in 2.2 without threads but instead 40 processes. I will also grab perf data from both 2.2 threaded and 2.2 without threads.

I think you don't even need to waste your time trying with 40 processes. There's something wrong going on there that we need to understand. Maybe it will be difficult to address, maybe it's stupid, maybe it could be worked around by a small config change. But anyway it's not normal to saturate 40 CPUs like this so I'm sure there's something causing a high contention.

Could you please also post, for 2.2-threads, the output of `show activity` on the CLI (will not disclose any confidential info) and of `show info` (will reveal your host name, feel free to hide it if you want) ?

Also, just a copy-paste of "perf top" running for 5 seconds or so would definitely help spot an entry point.

I'm a bit confused by the fact that the "top" output for 1.6 shows plenty of processes with unequal load while the original config shows "nbproc 2". Is it with a higher "nbproc" setting or maybe multiple generations of old and new processes ? What's your rough workload to use that many CPUs ? Mostly SSL maybe ? I'm asking because the stats for 2.2.2 shows 8k req/s and I see now way to use more than a few percent of a single CPU with just that so we really need to figure where the problem is.

I'd be interested in making you build with lock debugging enabled, but retrieving the stats output will require to run in foreground and I understand that it will not be easy in your environment. Thus I'd rather stick to easier solutions for now. I'll see if I can implement a "show lock" on the CLI and backport it instead.

**Comment by ngaugler** [comment]
Unfortunately the use case that existed when I first opened this issue no longer exists in our platform.  Perhaps I should have opened a new issue.  The use case that still exists is in the config files in https://github.com/haproxy/haproxy/issues/822#issuecomment-754867036

I attempted to gather data.  Unfortunately this will be heavy on SSL since we are doing only 1/5th the requests we are experiencing an abnormal use scenario with connections attempting to be established.

[haproxy-2.2-show.txt](https://github.com/haproxy/haproxy/files/5871725/haproxy-2.2-show.txt)
[haproxy-2.2-perf-top.txt](https://github.com/haproxy/haproxy/files/5871726/haproxy-2.2-perf-top.txt)
[haproxy-1.6-perf-top.txt](https://github.com/haproxy/haproxy/files/5871727/haproxy-1.6-perf-top.txt)
[haproxy-2.2-perf-report.txt](https://github.com/haproxy/haproxy/files/5871728/haproxy-2.2-perf-report.txt)

**Comment by wtarreau** [comment]
Don't be sorry. You can't decide what trouble will affect your workloads nor when, and this issue continues to match your initial problem.
Your "perf report" shows something extremely interesting, your process is spending all its time in the SSL lock. Your load is far from being exceptional, however the number of tasks in the run queue is very high, which indicates a saturation. I suspect that the SSL lock is grabbed when looking up TLS sessions, and that given there are likely a lot, the operations there are slower, making things worse. This remains strange because openssl uses multiple locks to reduce contention (the number of locks in use is returned by CRYPTO_num_locks(), I don't know how large it is by default, I'll run some tests). I'm suspecting that this has to do with false sharing of cache lines. I'll run some tests with one lock per cache line to see if that improves things or not.

**Comment by capflam** [comment]
I may be wrong, but looking at your 2.2 config, there is no thread mapping. It could help to set a cpu-map for your threads, just like for the process. You may add following line in your global config:
```
cpu-map auto:1/all   0-63
```
@wtarreau, it could help to use this kind of mapping by default. It is not always the best mapping but, by default, it is probably better that no mapping at all.

**Comment by ngaugler** [comment]
I tried to find a way to build debug with locks but could not.  If there's something else I can run please let me know.

**Comment by wtarreau** [comment]
@capflam I prefer not to set any mapping than to force one unless we know **exactly** how interrupts are bound. There's nothing worse than a userland thread bound to the same CPU as a ksoftirqd thread doing all the network job.

@ngaugler it's possible to enable thread debugging using -DDEBUG_THREAD but based on your perf report, I can already predict that it will indicate that all the time is spent in SSL_LOCK.

I've checked on my machine. With openssl 1.0.2, it uses 41 locks. I suspect they're not hashed but just set per family.
I'm attaching a patch that makes sure the locks are all aligned, to limit false sharing. However if one of them is reponsible for most of the work, it will not change much.

On my machine I'm doing 26k conn/s on 2 cores / 4 threads using this. I've not compared to wit
[0001-EXP-cache-align-SSL-locks.patch.txt](https://github.com/haproxy/haproxy/files/5872305/0001-EXP-cache-align-SSL-locks.patch.txt)
hout yet.

**Comment by ngaugler** [comment]
I compiled the patch into the 2.2.8 and built with `DEBUG=-DDEBUG_THREAD`.  Immediately after restarting it prints:  `locks=41`   It may be working.  After a minute or so of session rebuilding the CPU usage settles down, with CPU usage settling in  around 40% (we're only at 65K rps since its 4AM EST).   Although we normally require sessions to rebuild with haproxy 1.6 as well, since traffic is low haproxy settled down in 5 seconds where as the 2.2.8 debug took at least 60 seconds.  I am not sure if this is because of debug and/or perf running or not.  I am attaching data from the new build.

[haproxy-2.2-patch1-vv.txt](https://github.com/haproxy/haproxy/files/5872443/haproxy-2.2-patch1-vv.txt)
[haproxy-2.2-patch1-perf-top.txt](https://github.com/haproxy/haproxy/files/5872444/haproxy-2.2-patch1-perf-top.txt)
[haproxy-2.2-patch1-show.txt](https://github.com/haproxy/haproxy/files/5872445/haproxy-2.2-patch1-show.txt)
[haproxy-2.2-patch1-perf-report.txt](https://github.com/haproxy/haproxy/files/5872446/haproxy-2.2-patch1-perf-report.txt)

**Comment by wtarreau** [comment]
I've spotted something abnormal in your `show info`, the SSL cache doesn't seem to work as well as it should:

```
SslCacheLookups: 623379
SslCacheMisses: 618780
```

I think you're dealing with a very large number of clients which don't make frequent requests, and that the SSL cache is too small so keep enough of them in it.

Apparently I can reproduce this when using `hey` here (all other tools don't produce this, probably indicating that it doesn't support tickets and doesn't store contexts):
```
$ socat - /tmp/sock1  <<< "show info" | egrep 'CumSslConns|SslRate|SslFrontendKeyRate|SslFrontendSessionReuse_pct|SslCacheLookups|SslCacheMisses'
CumSslConns: 110651
SslRate: 2389
SslRateLimit: 0
MaxSslRate: 2564
SslFrontendKeyRate: 2383
SslFrontendSessionReuse_pct: 0
SslCacheLookups: 110017
SslCacheMisses: 110649
```

Could you please try to add this to your global section and check in "show info" if it changes anything ? :

```
tune.ssl.cachesize 1000000
```

It will use a bit more memory (the default is 20000, or about 4MB RAM, here it will take ~200 MB).

**Comment by ngaugler** [comment]
I checked my /etc/haproxy-2.2/haproxy.cfg (and the haproxy-2.2.cfg I uploaded https://github.com/haproxy/haproxy/issues/822#issuecomment-754867036).   I already have that tune line configured.  I speculate the sessions were timing out when we were out of CPU and that's why they were not being reused.  If you look at the latest data (https://github.com/haproxy/haproxy/issues/822#issuecomment-767420527) you can see the show info SSL session cache looks much better.

**Comment by wtarreau** [comment]
Ah sorry I missed this one (too many configs now I guess). Do you think the 600s lifetime could prevent enough reuse ? Maybe it would be worth trying to comment the entry and possibly double the cachesize to see if it changes anything to the reuse rate. The least key computation the better, even for your client's response time.

By the way, your first test results look encouraging, so it may really be that we were causing too much contention there. Even the cache ration looks better, while it should not have changed. So it's also possible that the better performance is directly related to a better reuse ration hence less crypto being done.

In openssl 1.1.0 and above, they completely removed these locks, now they're internal to openssl. On my machine the performance is lower with 1.1.1 than with 1.0.2 but it might be caused by different algos being used. But it's also possible that the cost of locking is much lower.

**Comment by ngaugler** [comment]
I restarted with 2.2.8 and for whatever reason the system never calms down.  It's still burning 100% of the CPU.  The `perf report` shows the `ssl_locking_function` as the culprit again.    I increased lifetime to 1800 and   I rebuilt without DEBUG and restarted without perf and its still having issues.  There has to be something else related to this that isn't resolved.   `perf top` shows the same  `ssl_locking_function` culprit.  I don't think the lifetime should be an impact right now since we're only running for 2minutes.    

EDIT - It finally calmed down.  It took almost 4 minutes.  That's troubling since restarting with 1.6 takes <10 seconds to settle down with this traffic level.  I wonder if it's really fixed or if traffic volume is just low enough that it is able to catch up where as during the day it cannot catch up, ever.

**Comment by wtarreau** [comment]
OK so that would mean that it's new connections that are costing much. It's been a while since I haven't seen cache entries because nowadays virtually all clients use tickets, and tickets have the nice property of surviving reloads and even node migration. Do you know if there's a particular reason your clients do not support tickets ? They are probably very specific to that task, which I could obviously understand.

I'll see if I can improve the locking code to report stats on the function name an line number (which openssl already provides to the locking function). This will mostly tell us the number of passes, but could help us figure what it's doing that slowly with a lock held.

If in the mean time you have the opportunity to try to build against openssl-1.1.1 that could be great. In our INSTALL file there's even a howto to help on this.

**Comment by ngaugler** [comment]
I just confirmed the haproxy 2.2.6 without the patch settles down after 4minutes as well.  Unfortunately many of the connections on this server are server to server connections and I presume many of the companies are running on legacy software.  I will work on building with openssl 1.1.1i right now.  I joined the Haproxy Slack, not sure if that would be easier?

**Comment by ngaugler** [comment]
Ok - I compiled with openssl 1.1.1i.  This version started up and showed a different problem in `perf top`, but only used ~90% CPU.   It still took 3 minutes to settle down before CPU utilization was around 40%, but it was faster than the other 2.2 versions.  Still much slower than 1.6 which settles down in a few seconds.  Attached data.

[haproxy-2.2-openssl-vv.txt](https://github.com/haproxy/haproxy/files/5872678/haproxy-2.2-openssl-vv.txt)
[haproxy-2.2-openssl-perf-top.txt](https://github.com/haproxy/haproxy/files/5872679/haproxy-2.2-openssl-perf-top.txt)
[haproxy-2.2-openssl-show.txt](https://github.com/haproxy/haproxy/files/5872680/haproxy-2.2-openssl-show.txt)

**Comment by ngaugler** [comment]
Also, once it settles down the main culprit in perf appears to be `chash_get_server_hash`.  Not sure if we have another problem there.  Example:
```Samples: 406K of event 'cycles:pp', Event count (approx.): 1271827230124598
Overhead  Shared Object              Symbol
   8.38%  haproxy-2.2.8-openssl      [.] chash_get_server_hash
   3.24%  haproxy-2.2.8-openssl      [.] __bn_sqrx8x_reduction
   2.93%  haproxy-2.2.8-openssl      [.] conn_backend_get
   2.69%  [kernel]                   [k] mds_user_clear_buffers
```

**Comment by ngaugler** [comment]
I restarted with haproxy 1.6.   Everything calmed down in less than 5 seconds.  I checked the 'show info' on one of the processes and the SslCacheLookups was about 1.5x the SslCacheMisses.  So the cache ratio appears to be just as bad on 1.6, but for whatever reason it's much more efficient at handling new sessions. 

Also, for what it's worth, once 1.6 settled down CPU usage was much less than 2.2.    27% instead of 40%.  So the ongoing problem of high CPU usage under normal operation could be a completely different issue than this start up issue.  I added a graph.  The red arrow is where  I restarted with haproxy-2.2.8-openssl.  You can see how long the CPU is abused before it settles down around 40% utilization.  The blue arrow is where I restarted with haproxy 1.6.  The 5 seconds of increased CPU utilization doesn't even show up and then it settles down to 27% utilization.  I fear we have at least two separate issues.

![cpu-usage](https://user-images.githubusercontent.com/1454960/105834309-eb71d500-5f8f-11eb-95f4-43fd1d7fad79.PNG)

**Comment by wtarreau** [comment]
The perf top shows that now most of the work happens in low level crypto stuff. I'm wondering if anything changed regarding this between 1.6 and 2.2. For example I remember we changed the default dhparam to 2k but you were already setting it in 1.6.

I can understand how the overall idle is reduced from 1.6 to 2.2, due to all the architecture changes that went in to support more flexible layering and threads that I mentioned previously. But that doesn't explain why it takes so long to settle on new connections upon restart. Could you please share the "show info" on 1.6 so that we can see the SSL cache lookups/misses ? If for any reason something is broken there, it could explain that it takes ages to start, or that there is a constant background crypto noise.

**Comment by ngaugler** [comment]
I am attaching a `show info` from one of the frontend processes. I believe my 1.6 config has 30 frontend processes.
[haproxy-1.6-show-info.txt](https://github.com/haproxy/haproxy/files/5873752/haproxy-1.6-show-info.txt)

**Comment by wtarreau** [comment]
So the reuse ratio is roughly the same there. I'm wondering now if anything could cause clients to negotiate a more intensive set of ciphers or TLS version, because I don't understand how it's possible that we're spending that much time doing crypto stuff only in one version. I'm still checking what exactly we're having in the runqueue for now.

**Comment by ngaugler** [comment]
It looks like I redacted the SSL cipher configuration, I am not sure why.  We use the same list of ciphers in 1.6 as we do 2.2.  Since the original 2.2 test ran with the same version of ssl as 1.6, I presume it would use the same ciphers.  Unless there were changes in the haproxy code.

`ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA:ECDHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA:AES256-SHA:DES-CBC3-SHA:!kEDH:!LOW:!EXP:!MD5:!aNULL:!eNULL`

After sleeping on the issues, I was thinking.  This implementation is predominantly server-to-server with very limited client-to-server requests.  This means that each new HTTPS connection will most likely have a new SSL session and will exist for several minutes, before closing and never reusing the same SSL session.  I know that https connections will exist for at least 2.5 minutes because we had to increase `timeout http-keep-alive` to 150 seconds to make Amazon happy.  Now the new SSL creation problem we are experiencing cannot be linked to long sessions since when restarting 1.6 takes 5 seconds to establish all of the connections where as 2.2 takes 240 seconds, or 48x the CPU.


I still believe we have at least a second issue.  Under normal operation last night 1.6 utilized only 27% CPU where as 2.2.8 patched used 40%.  I understand there will be extra CPU usage under 2.2, but I strongly feel that ~50% more (40-27)/27= 48 is excessive.  The 50% number I saw last night was mostly on par with the original issue (https://github.com/haproxy/haproxy/issues/822#issuecomment-682228178).  I do not think we should fight battles on multiple fronts at this time unless we split this issue into multiple issues.

**Comment by wtarreau** [comment]
I agree that *in theory* there shouldn't be differences in ciphers etc. But as I said, I remembered that there were some changes (such as dhparam) that could have had an impact. I'm still quite bothered by this 48x stabilization time. It's possible that we're really facing a 48x cost for whatever reason, or that some operations are failing and being retried. The 240 seconds you're observing is not that much higher than the average session duration you observed and we could imagine that it needs all sessions to be renewed to become stable (why, I don't know). The CPU issue should be left aside for now, unless it helps explaining the issue. I mean, if for whatever reason we're facing a much higher cost for a new session, it's expected that on the long run a significant portion of the CPU usage will come from this. And conversely, if there's an accumulation effect that slows things down, it would be possible that clients give up and try again, resulting in accumulating plenty of handshakes that increase the resync time. I'm totally amazed by the size of the runqueue in 2.2. For example you have roughly 98k conn in 1.6 and 138k in 2.2. If the 40% difference corresponds to failed handshakes that are being retried, this can induce a significant cost. That's why I'm still searching on my side what all these connections are doing. I can reproduce a large run queue too (with much lower numbers) and I'd like at least to explain this.

**Comment by ngaugler** [comment]
I decided to dig through the code earlier to try and figure out why functions like `chash_get_server_hash` showed up in perf.  I couldn't find anything in the code that seemed like a cause.  I started thinking about something you suggested earlier, what if it's threading?  So I decided to compile 2.2.8, clean with the patch, without debugging, without openssl 1.1.1 and without threading (USE_THREADS=).  I copied my configs for 1.6 into a new directory.  After a small change to move arguments from `options httpchk` to `http-check send` the configs were valid.  I just restarted and immediately 2.2.8-nothreads was able to recover all of the sessions and serve traffic.   CPU utilization still isn't on par with 1.6.  Before the restart CPU utilization was around 35%.  Now it's at 55%.  So there's still problems with 2.2.8 that cause significantly more CPU usage.  But at least I wasn't losing CPU cycles to `ssl_locking_function` or `chash_get_server_hash`.  Now the biggest CPU usage was kernel `mds_user_clear_buffers` (just like 1.6), followed by `__bn_sqrx8x_reduction`.

I attached a graph showing CPU usage.  The red line being when I restarted with haproxy-2.2.8-nothreads and the blue line being when I restarted with haproxy 1.6


![cpu-usage](https://user-images.githubusercontent.com/1454960/105950397-f2522380-6033-11eb-8384-d53b48d24ae9.PNG)

**Comment by wtarreau** [comment]
Hi Nicholas,

thanks for your detailed report and experiments. At this point, not only did I manage to reproduce your restart problem, but I could explain it and address it. The root cause is that connections are accepted faster than they can be processed, and this is what resulted is this huge runqueue that bothered me. With only 10k clients I managed to reach 350k tasks in the runqueue and something like 180k FDs. The problem was that we were swallowing incoming traffic without even knowing if it was possible to digest it. And SSL handshakes are a typical amplifier of this. In older versions, the operations were much more sequential so it was not much possible to accept that many connections without having to deal with their handshakes.

I intend to merge the attached patch into 2.4 and to backport it to older versions (2.3 and 2.2, probably not other ones due to more important changes there). I'd appreciate it if you could give it a try. Then I'll have a look at this chash stuff, it seems to remind me something related to server maxconn, but I need to check before saying stupidities.

[0001-BUG-MEDIUM-listener-do-not-accept-connections-faster.patch.txt](https://github.com/haproxy/haproxy/files/5881714/0001-BUG-MEDIUM-listener-do-not-accept-connections-faster.patch.txt)

I wanted to add, the patch above was already a backport to 2.2 and it will only apply to 2.2 (and it was tested there as well).

**Comment by ngaugler** [comment]
Unfortunately I do not feel this fixes the problem.  If I restart with 1.6 it recovers in <3 seconds.  I am still out of CPU almost 2 minutes later.

```   8.76%  haproxy-2.2.8-patch2  [.] ssl_locking_function                                                                                                                                                                                                                                                                    ▒
   4.58%  [kernel]              [k] __slab_free                                                                                                                                                                                                                                                                             ▒
   4.09%  libcrypto.so.1.0.0    [.] 0x00000000000c882c                                                                                                                                                                                                                                                                      ▒
   3.96%  [kernel]              [k] default_send_IPI_mask_sequence_phys                                                                                                                                                                                                                                                     ▒
   2.78%  libcrypto.so.1.0.0    [.] 0x00000000000c867c                                                                                                                                                                                                                                                                      ▒
   2.62%  libcrypto.so.1.0.0    [.] 0x00000000000c90ab                                                                                                                                                                                                                                                                      ▒
   2.62%  libcrypto.so.1.0.0    [.] 0x00000000000c922e                                                                                                                                                                                                                                                                      ▒
   2.62%  [kernel]              [k] __dentry_kill                                                                                                                                                                                                                                                                           ▒
   2.62%  libcrypto.so.1.0.0    [.] 0x00000000000c84a1                                                                                                                                                                                                                                                                      ▒
   2.46%  libcrypto.so.1.0.0    [.] 0x00000000000c90a5                                                                                                                                                                                                                                                                      ▒
   2.46%  [kernel]              [k] __netdev_pick_tx                                                                                                                                                                                                                                                                        ▒
   2.46%  libcrypto.so.1.0.0    [.] 0x00000000000c8ba0                                                                                                                                                                                                                                                                      ▒
   2.29%  libcrypto.so.1.0.0    [.] 0x00000000000c9183                                                                                                                                                                                                                                                                      ▒
   2.29%  [kernel]              [k] smp_call_function_many                                         ```

**Comment by ngaugler** [comment]
This patch appears to correct the issues as you discussed.  Unfortunately it is too difficult for me to tell if it is as good as 1.6, but it definitely makes it so 100% cpu utilization does not occur for an extended period of time.  I think what you identified as a problem is definitely the cause.  Over the years I have fine tuned the `backlog` settings to ensure that the individual processes were not overloaded during a restart, so it seems to be on target.  The good news is the system is able to handle all of the traffic at 2AM EST.  The bad news is we are using 60% of the CPU instead of 40% of the CPU (increase of 50%).  I have attached a copy of perf top.  Please note, I am running both patches you proposed in this issue in this test. `chash_get_server_hash` is definitely the next place to look.

[haproxy-2.2-patch2-perf-top.txt](https://github.com/haproxy/haproxy/files/5885163/haproxy-2.2-patch2-perf-top.txt)

**Comment by wtarreau** [comment]
Great, thanks for the test. Indeed, next one definitely is chash_get_server_hash(). There are two factors which inflate its cost:
  - the number of draws for the random call (default is 2, if you use `balance random(1)` it should be halved)
  - the redispatch on connection failure, which tries hard to pick a different server. *this* is expensive as it requires to repeat the draw until we're on a different one. As I'm seeing retry 0 in your config, I think this does not happen, but I could be wrong.

**Comment by ngaugler** [comment]
I am open to suggestions on what config changes you would prefer.  Please let me know how to proceed.

**Comment by wtarreau** [comment]
The easy change you can do and which doesn't require to wait for the peak hour is to use `balance random(1)` instead of `balance random`. It should immediately half the cost of `chash_get_server_hash`. I'm still looking how we could get a lower cost in this function and I must confess I'm surprised that it costs that much for you. One feature that's known for increasing its cost is the `hash-balance-factor` directive which is used to pick the least loaded server, but you don't have it in your config. One possible factor playing a bit here is the extra tree depth coming from the weights that cause extra nodes to be stored. If you only use weights 100 and 150 you could for example use 10 and 15 and reduce by ~7 the number of levels in the tree. But since there doesn't seem to be retries in your case and no hash-balance-factor, there is a single lookup in the tree so there should be no contention there.

**Comment by wtarreau** [comment]
Wait a minute, in 2.2 the chash tree was protected using a spinlock, that turned to an rwlock in 2.3. For course with 40 threads this leads to a huge contention.

Are you interested in testing 2.3 ? Alternately I can see if the patches to turn that to an rwlock are suitable to 2.2 and provide them to you so that you apply them to your tree.

**Comment by ngaugler** [comment]
I am willing to test whatever will help us advance.  Are the patches that you proposed in this issue in 2.3?

**Comment by wtarreau** [comment]
Yes definitely, the change was made between 2.2 and 2.3.
I've run some tests here, with only 4 threads (hence little contention, mostly lookup cost)
  - balance random(1) is ~5% faster than random(2) which is the default
  - weight 10/15 are about 6% faster than weight 100/150 for random
  - leastconn is always faster than random here (remember, in 2.2, lock contention was the killer)

Leastconn has one known corner case where it can be more expensive, which is when maxconn is reached, but you don't have a maxconn on your servers.

I'm attaching 4 patches backported from 2.3 to 2.2 that improve the locking there for 3 algos: `chash` (all consistent hashes including random), `first`, `leastconn`. You don't even need `first`, it's just because it's in the series, but feel free to drop it. It will not be as advanced as in 2.3 but just with this with 4 threads, chash_get_server_hash() which was at 2% in my tests totally disappeared from `perf top`. If that's easier to you, 2.3 is an option. Note that we'll probably issue another one today with a few more fixes.

I think that sticking to `balance random` remains a good option for you. If you want to optimize further, leastconn is better, but keep in mind that it can have corner cases, and that updating the server comes with a write lock that may balance the other savings with 40 threads.

[0004-MINOR-lb-leastconn-only-take-a-read-lock-in-fwlc_get.patch.txt](https://github.com/haproxy/haproxy/files/5885749/0004-MINOR-lb-leastconn-only-take-a-read-lock-in-fwlc_get.patch.txt)
[0003-MINOR-lb-first-use-a-read-lock-in-fas_get_next_serve.patch.txt](https://github.com/haproxy/haproxy/files/5885750/0003-MINOR-lb-first-use-a-read-lock-in-fas_get_next_serve.patch.txt)
[0002-MINOR-lb-chash-use-a-read-lock-in-chash_get_server_h.patch.txt](https://github.com/haproxy/haproxy/files/5885751/0002-MINOR-lb-chash-use-a-read-lock-in-chash_get_server_h.patch.txt)
[0001-MINOR-backend-replace-the-lbprm-lock-with-an-rwlock.patch.txt](https://github.com/haproxy/haproxy/files/5885752/0001-MINOR-backend-replace-the-lbprm-lock-with-an-rwlock.patch.txt)

**Comment by wtarreau** [comment]
I'm going to revert the first patch (that regulates connection acceptance based on the load). It's causing lots of trouble in presence of high-speed traffic, as for a reason I cannot explain, the queues are constantly full, just as if some tasks were always woken up before knowing if they can work :-(

The net effect is that once you're transferring at 50-60 Gbps, you cannot connect anymore until a connection is released. This is even valid for the stats socket (which is how I detected it). So we have a deeper chicken-and-egg issue that we need to sort out. In your case with small responses it's not an issue but it definitely is for many other use cases, and I cannot even afford to backport it.

**Comment by ngaugler** [comment]
I downloaded the source for 2.3.4, applied only the second patch, and compiled with openssl-1.1.1.  Patch rejected hunk 2, but it was simple and only added `, queued++` in the appropriate place in the for loop, so I manually applied it.   I have not yet tested this on my primary server, traffic is too busy during this time of day.  I tested haproxy-2.3.4-patch2-nothreads with no threads on a different implementation.  CPU usage jumped from 25% to 36%.  I will try and test the nothreads and threads version of this on the main server tonight or tomorrow night.  Attached is the perf top from this alternate implementation. 

The only change I made to the config file was to add:  `nbthread 1`   I realize this is only for the non-threaded version.

[haproxy-2.3.4-patch2-nothreads-alternate-perf-top.txt](https://github.com/haproxy/haproxy/files/5907448/haproxy-2.3.4-patch2-alternate-perf-top.txt)

**Comment by wtarreau** [comment]
I'll propose something better soon. I could verify that we're indeed performing the handshakes of pending closed connections and that this is what is ruinning the performance and making the situation worse. Just an example here with an accepted connection that was reset in the system's accept queue but the error is only detected while sending, after the handshake had consumed one ms of CPU:
```
12:50:48.236056 accept4(8, {sa_family=AF_INET, sin_port=htons(62794), sin_addr=inet_addr("127.0.0.1")}, [128->16], SOCK_NONBLOCK) = 1109
12:50:48.236071 setsockopt(1109, SOL_TCP, TCP_NODELAY, [1], 4) = 0
12:50:48.257270 getsockopt(1109, SOL_SOCKET, SO_ERROR, [ECONNRESET], [4]) = 0
12:50:48.257281 write(-1, "ABORT conn=0x3c2dc00 fd=1109 flg=0x8805300\n", 43) = -1 EBADF (Bad file descriptor)
12:50:48.257297 recvfrom(1109, "\26\3\1\2\0", 5, 0, NULL, NULL) = 5
12:50:48.257310 recvfrom(1109, "\1\0\1\374\3\3\267\356`l\2\250y\316\0a\272Q\n\237?\355\252&\375\343\20\254\311{.8E\316v;p\r z\212\200\337\f0yy\334A\2$\237\301\224jR\31\236d\25\211s\240\31\245\250(\342(i\353\0>\23\2\23\3\23\1\300,\3000\0\237\314\251\314\250\314\252\300+\300/\0\236\300$\300(\0k\300#\300'\0g\300\n\300\24\09\300\t\300\23\0003\0\235\0\234\0=\0<\0005\0/\0\377\1\0\1u\0\v\0\4\3\0\1\2\0\n\0\f\0\n\0\35\0\27\0\36\0\31\0\30\0#\0\0\0\26\0\0\0\27\0\0\0\r\0000\0.\4\3\5\3\6\3\10\7\10\10\10\t\10\n\10\v\10\4\10"..., 512, 0, NULL, NULL) = 512
12:50:48.258004 sendto(1109, "\26\3\3\0z\2\0\0v\3\3\322\262?\361\367\244\257\201\345\314J<\212\213\2166L\354\213\345\371\266\270\1\326\251\202\354u\356I\360 z\212\200\337\f0yy\334A\2$\237\301\224jR\31\236d\25\211s\240\31\245\250(\342(i\353\23\2\0\0.\0+\0\2\3\4\0003\0$\0\35\0 -\32\273E\262d\326\260]\r}\331XPH%\"\364\237D\227\331\311\246\5p\nY\362i\274E\24\3\3\0\1\1\27\3\3\0\27\37@\24y\250DF\371\241\32R\4\3\10\351O\323O\330\373\320\234\361\27\3\3\3\177\302\357\236\32\276\274\351\3\36\177m\4\251T\243\37E\337\217=}\352)&\253\5\322\330\17\222\372s\272\340"..., 1421, MSG_DONTWAIT|MSG_NOSIGNAL, NULL, 0) = -1 EPIPE (Broken pipe)
12:50:48.258036 close(1109)             = 0
```
As you can see, the test on the socket error already reported a pending error before starting the operation. I could verify that by doing this prior to the SSL handshake and aborting, I can totally fix the problem and restore the old behavior. I'm just trying to figure the cleanest way to do it, as we don't want to lose pending server responses either.

**Comment by wtarreau** [comment]
Note, above I filtered only on fd 1109 which is why there's such a huge delay between accept() and entering the SSL handshake. Howevre between the last recvfrom and sendto, there was nothing but SSL calculations.

**Comment by wtarreau** [comment]
So I'm quite puzzled, because in 2.4-dev we're accepting this through early data, despite allow-0rtt not being in the config!

**Comment by wtarreau** [comment]
Anyway, now I think I understand the major difference between 1.6 and 2.x here. In 1.6 since the handshake started immediately after accept(), this would slow down accept() and make excess connections expire in the system's queue. As such there were few unneeded handshakes. With 2.x and the accept queue, many of them are accepted and expire later, while waiting for previous ones to compute their handshake. As such they're already in the queue and will be calculated anyway, explaining the huge load and persistent lag that causes future connections to be delayed in turn and to fail.

**Comment by wtarreau** [comment]
Here comes a tested patch. For me the difference is huge because after stopping the traffic it recovers in 2 seconds instead of 1 minute, and this makes more sense than playing ugly tricks with the runqueue (with this said I'm still thinking about moderating the accept batch length based on load now). I'm attaching the 2.4, 2.3 and 2.2 versions which I backported to make your life easier with them. I'm going to merge the 2.4 one anyway since it *does* address a real issue. I wanted to issue 2.2 and 2.3 last week but did postpone them due to this troubling problem, so I intend to backport these very soon. Please just let me know if you prefer to test early or to just apply the next update. Anyway here come the patches.

[2.3-BUG-MEDIUM-ssl-check-a-connection-s-status-before-co.patch.txt](https://github.com/haproxy/haproxy/files/5912163/2.3-BUG-MEDIUM-ssl-check-a-connection-s-status-before-co.patch.txt)
[2.2-BUG-MEDIUM-ssl-check-a-connection-s-status-before-co.patch.txt](https://github.com/haproxy/haproxy/files/5912164/2.2-BUG-MEDIUM-ssl-check-a-connection-s-status-before-co.patch.txt)
[2.4-BUG-MEDIUM-ssl-check-a-connection-s-status-before-co.patch.txt](https://github.com/haproxy/haproxy/files/5912165/2.4-BUG-MEDIUM-ssl-check-a-connection-s-status-before-co.patch.txt)

**Comment by ngaugler** [comment]
I will attempt to run some tests on the alternate platforms today with 2.2 and 2.3.   I cannot run it in the large scale environment until after midnight and it is too far away to know what the night will bring for availability.  Just to confirm, I am going to run both `0001-BUG-MEDIUM-listener-do-not-accept-connections-faster.patch.txt` and `BUG-MEDIUM-ssl-check-a-connection-s-status-before-co.patch.txt`.

**Comment by wtarreau** [comment]
No, please do not even use the "do not accept connections faster". Just the latest one. It must be enough in that it will bring back the behavior closest to the 1.6 era.

**Comment by ngaugler** [comment]
In the alternate environment both 2.2 and 2.3 perform well without threads.  1.6 was using ~27% CPU, 2.2 used ~30% and 2.3 used ~32%.  It will be interesting to see how this performs on larger scale tonight without threads and with threads.

**Comment by wtarreau** [comment]
Great, thanks. But please do not forget to get some sleep, you've spent many nights upgrading/downgrading it!

**Comment by hloeung** [comment]
I've been following along and have been really impressed with the time and effort both @ngaugler and @wtarreau have put in with this. Thanks to both of you for making HAProxy that better.

Now, if only @ngaugler could reproduce the bug we hit here https://github.com/haproxy/haproxy/issues/674 :laughing: 

Either way, can't wait to try out the latest version with all these new performance improvements.

**Comment by wtarreau** [comment]
@hloeung you should retry the latest 2.4-dev (I wanted to issue another one this week-end but was kept busy trying to figure what was wrong with this issue). We've added more observability tools in 2.4 to further help narrow performance problems down. Some important scheduler improvements were made over 2.2 as well, with traffic classes which help keep a very low latency in front of I/O storms. Regardless, better update the thread there with your findings once you retest.

**Comment by ngaugler** [comment]
Ok for the tests tonight I reset the source code to that of the releases.  I then patched with the single patch file for src/ssl_sock.c with two hunks and compiled with openssl 1.0.2g (this was a mistake, I meant to use 1.1.1i)  and USE_THREAD=.  For 2.x I ran `balance random` and for 1.6 we continued with the `balance leastconn`  Don't fix what isn't broken there.

Restarting with 1.6, 2.2, and 2.3 seemed to stablize the CPU usage in about 5 seconds.  However, it appears some of our client connections do not restart right away and run on a delay, maybe 10 sec after restart and another round 20 seconds after restart.  On 1.6 and 2.3 this caused a few processes to jump back up, but quickly fall down in 1 second.  On 2.2 this did not appear to work as smoothly.  Many processed jumped up, and took more than 1 second to recover, but did recover after a few seconds.

CPU Utilization once settled down:
1.6  30%
2.2  47%
2.3  46%

I am still alarmed that both 2.2 and 2.3 utilize 50% more CPU than 1.6.  I have not begun to test with threading compiled, but will do so in a bit.

[haproxy-2.2.8-patched-nothreads.txt](https://github.com/haproxy/haproxy/files/5916533/haproxy-2.2.8-patched-nothreads.txt)
[haproxy-2.3.4-patched-nothreads.txt](https://github.com/haproxy/haproxy/files/5916532/haproxy-2.3.4-patched-nothreads.txt)


EDIT I made the mistake of compiling with the old version of SSL.  I quickly ran another test with haproxy 2.3 with openssl 1.1.1i under the same circumstances and attached a new file with debug info.

[haproxy-2.3.4-patched-nothreads-openssl111.txt](https://github.com/haproxy/haproxy/files/5916625/haproxy-2.3.4-patched-nothreads-openssl111.txt)

**Comment by ngaugler** [comment]
For the threading test, I kept the same as above: clean source, single patch for src/ssl_sock.c, openSSL 1.1.1i,  `balance random`.  I did not configure nbproc/nbthread and let the binary decide to create 40 threads.     It's later in the night, so 1.6 was using only 21% before I started this test.    This did not go well at all.  It took over 1 minute (68 seconds to be exact) for the CPU usage to drop from 80-100% to a stable point of 35%.   I am attaching data from when the CPU was high.  I do not think we need to grab perf data from once it is stable.  

[haproxy-2.3.4-patched-threads.txt](https://github.com/haproxy/haproxy/files/5916684/haproxy-2.3.4-patched-threads.txt)


Just to cover all of my angles, I ran this same binary compiled with thread support with the configurations for no threads used in the previous comment.   I saw mostly consistent results as running a binary compiled with USE_THREAD=.

**Comment by wtarreau** [comment]
Now I'm totally puzzled and have no idea what to look for anymore :-(

Basically we're seeing that this machine is saturated with 5.8k SSL handshakes per second and that it's still spending 34% of its time doing SSL keys. So the legitimate question now is: how is it possible that one version achieves in 5 seconds what the other one achieves in 68 seconds, 23 seconds of which are supposed to be exactly the same (SSL key calculation) ? I only foresee three possibilities to explain this:
  - either for a reason we can't figure, the crypto code is 5-10 times more expensive when called by recent versions, but I think it doesn't really make sense ;
  - or we're still doing way too many handshakes and still delivring them too late for the clients, which expire and try again (I guess your users use very short timeouts).
  - or the impatient clients close "too cleanly" so that getsockopt(SO_ERROR) cannot detect their closure. But with the SSL stack reading exactly the number of expected bytes we can't detect a possible close. I'm seeing how I could hack something like this using recv(MSG_PEEK). A cleaner solution would be to have the transport layer implement its own buffer and try to read again after delivering data back to the stack. That could be enough to detect errors and closures.

Sadly we don't have all the profiling tools in 2.3 that we have in 2.4, and some of them are really useful as we can measure each scheduled task's experienced latency. I can try to further backport some of them but I know that some are a bit tricky so I won't take risks. On the other hand, the modified areas have probably not changed between 2.3 and 2.4.

Another point, in my tests I noticed we were passing through SSL_read_early_data() while there's no 0-rtt here. I need to have a closer look into this as it doesn't seem to make sense, either indicating we have a bug or that I'm completely overlooking something.

If you're interested, I can implement a quick patch to try to detect if it's just SO_ERROR that doesn't catch all closures, make sure we also detect clean ones, and add counters observable in "show activity" to help spot this.

**Comment by wtarreau** [comment]
I think that for the unexpected early data calls, we need this patch. I think @cognet will agree. I don't think it will change anything to the performance issue you're observing, though.

I straced both 2.4 and 1.6 and could confirm that the native FD code in openssl doesn't handle pending errors either so this alone is not the cause and rather an amplifier. I could modify ha_ssl_read() to perform large reads at once into a buffer and make sure to detect pending ends (read shut or errors). However since the calling stack doesn't care, we'll instead have to preload pending data before starting the handshake so that we have a chance to run the error check that the stack ignores. But it still leaves the problem of why we're seeing that behavior in the first place :-/

[0001-WIP-BUG-MINOR-ssl-do-not-try-to-use-early-data-if-no.patch.txt](https://github.com/haproxy/haproxy/files/5917727/0001-WIP-BUG-MINOR-ssl-do-not-try-to-use-early-data-if-no.patch.txt)

**Comment by ngaugler** [comment]
This afternoon I spent the time to setup an environment to test in another facility that is almost similar in use case to the large scale environment, but with significantly less traffic. This will allow me to do more testing during the day.  The patch you provided would not build on 2.3.  I am attaching a patch file that I used to build on 2.3.  I have not tested 2.2.   I feel it would be advantageous for us to simplify this debugging process and attempt to narrow this problem down in one version.  Once we solve it there it can be ported to the other versions if applicable.  My immediate objective is to procure hardware on a 4 year EOL cycle.  That being said I am OK with the version we choose to resolve these issues being 2.4 since that version is the version of the future.  I leave the decision on which version to prioritize to you.

[2.3-0001-WIP-BUG-MINOR-ssl-do-not-try-to-use-early-data-if-no.patch.txt](https://github.com/haproxy/haproxy/files/5922264/0001-WIP-BUG-MINOR-ssl-do-not-try-to-use-early-data-if-no.patch.txt)

I will test this patch in the large scale environment tonight.


Attached are the data points from the new environment. I am not sure if this helps or is just muddying the water.  Please note this environment has two threads `cpu-map auto:1/39-40 0 20` which do not handle any SSL and have a different use case.  The standard use case is on `cpu-map auto:1/1-38 1-19 21-39`

[haproxy-23-nothreads.txt](https://github.com/haproxy/haproxy/files/5922627/haproxy-23-nothreads.txt)
[haproxy-23-threads-settling.txt](https://github.com/haproxy/haproxy/files/5922628/haproxy-23-threads-settling.txt)
[haproxy-23-threads-settled-down.txt](https://github.com/haproxy/haproxy/files/5922629/haproxy-23-threads-settled-down.txt)

**Comment by ngaugler** [comment]
I decided to start an hour earlier to ensure I had sufficient traffic volume.  Unfortunately even with the `2.3-0001-WIP-BUG-MINOR-ssl-do-not-try-to-use-early-data-if-no.patch` applied on top of `2.3-BUG-MEDIUM-ssl-check-a-connection-s-status-before-co.patch` the result was still the same.  nbproc 40 settled down shortly after (not as fast as 1.6 though) and the 40 threaded version did not settle down. I had to pull it back since the number of served req/sec fell off by over 70%.

[haproxy-2.3.4-patched-nothreads.txt](https://github.com/haproxy/haproxy/files/5923299/haproxy-2.3.4-patched-nothreads.txt)
[haproxy-2.3.4-patched-threads.txt](https://github.com/haproxy/haproxy/files/5923300/haproxy-2.3.4-patched-threads.txt)

I will hang around for a few hours if you want to supply some quick patches.

**Comment by wtarreau** [comment]
Hi @ngaugler, please try again on 2.4-dev starting with commit 76390dac0. This series improved the request rate by up to 25% compared to 2.4-dev9 and has divided the long-tail request latency by 10 in presence of heavy TLS handshakes (and even by 20 before 99.9%). If you're still observing issues, I'll be interested in you building with `-DDEBUG_TASK`, adding `profiling.tasks on` in your global section, and issuing `show tasks` and `show profiling` on the CLI under load to figure who's causing these delays.

**Comment by wtarreau** [comment]
FWIW 2.4-dev10 is now out with all this in it, so no patch needed.

**Comment by wtarreau** [comment]
And now dev11 with plenty of scalability fixes for other users following this issue.

**Comment by capflam** [comment]
I'm closing this issue. Feel free to reopen it if necessary. Thanks !
